TacTok/models/bpe.py

```diff
@@ -0,0 +1,118 @@
+
+from collections import Counter
+from typing import List, Dict, Tuple, Optional
+import typing
+
+
+class IndexedTrieNode:
+    def __init__(self, char: str) -> None:
+        self.char = char
+        self.is_end = False
+        self.index: Optional[int] = None
+        self.children: Dict[str, IndexedTrieNode] = {}
+
+
+class IndexedTrie:
+    def __init__(self) -> None:
+        self.root = IndexedTrieNode("")
+    
+    def insert(self, idx: int, token: str) -> None:
+        cur_node = self.root
+        for char in token:
+            if char in cur_node.children:
+                cur_node = cur_node.children[char]
+            else:
+                new_node = IndexedTrieNode(char)
+                cur_node.children[char] = new_node
+                cur_node = new_node
+        cur_node.is_end = True
+        cur_node.index = idx
+
+    def longest_prefix(self, sentence: str) -> Optional[Tuple[int, str]]:
+        cur_node = self.root
+        longest_match: Optional[Tuple[int, str]] = None
+        for idx, char in enumerate(sentence):
+            if cur_node.is_end:
+                assert cur_node.index is not None
+                longest_match = (cur_node.index, sentence[:idx])
+            if char in cur_node.children:
+                cur_node = cur_node.children[char]
+            else:
+                return longest_match
+        if cur_node.is_end:
+            assert cur_node.index is not None
+            longest_match = (cur_node.index, sentence)
+        return longest_match
+
+class LongestMatchTokenizer:
+    def __init__(self, vocab: List[str], include_unks: bool = False) -> None:
+        self.word_chunk_cache: Dict[str, List[Tuple[int, str]]] = {}
+        self.vocab = vocab
+        self.include_unks = include_unks
+        self.vocab_size = len(vocab) \
+          + (1 if include_unks else 0)
+        self.unk_index = len(vocab)
+        self.tok_trie = IndexedTrie()
+        for idx, word in enumerate(vocab):
+            self.tok_trie.insert(idx, word)
+
+    def pre_tokenize(self, sentence: str) -> List[str]:
+        return sentence.split(" ")
+
+    def tokenize_to_idx(self, sentence: str) -> List[int]:
+        return [chunk_idx for chunk_idx, chunk_str in
+                self.tokenize(sentence)]
+    def tokenize(self, sentence: str) -> List[Tuple[int, str]]:
+        words = self.pre_tokenize(sentence)
+        tokens = []
+        for word in words:
+            if word in self.word_chunk_cache:
+                tokens.extend(self.word_chunk_cache[word])
+            else:
+                rest_chars = word
+                word_tokens: List[Tuple[int, str]] = []
+                while len(rest_chars) > 0:
+                    prefix_pair = self.tok_trie.longest_prefix(rest_chars)
+                    if prefix_pair is None:
+                        if self.include_unks:
+                            word_tokens.append((self.unk_index, "<unk>"))
```

Comment:
What happens here when we have multiple different <unk>s now?

Reply:
The longest match tokenizer is agnostic to how you create the vocab; if we're talking about simply using all the words files for BPE, then either you pass in the merged counts as the global counts file when creating bpe, or load the counts separately and merge them with Counter.update(). If we're talking about the more complicated "add the locals tokens when embedding local vars" that we talked about, that's not implemented yet (future pull request), but it would be done by creating two LongestMatchTokenizers, and giving the global one a vocabulary from multiple calls to get_bpe_vocab.